torch/csrc/jit/test_jit.cpp

```diff
@@ -496,7 +435,168 @@ void interpStageTest() {
+  std::shared_ptr<tracer::TracingState> state;
+  variable_list trace_vars_in;
+  std::tie(state, trace_vars_in) = tracer::enter(fmap<tracer::TraceInput>(vars_in), 1);
+  auto trace_vars_out = test(trace_vars_in);
+  tracer::exit(trace_vars_out);
+  return state->graph;
+}
+
+variable_list grad(const variable_list& outputs, const variable_list& inputs, const variable_list& grad_outputs) {
+  static const auto get_edge = [](const Variable& v) -> edge_type {
+    return std::make_pair(v.grad_fn() ? v.grad_fn() : v.grad_accumulator(), v.output_nr());
+  };
+  auto & engine = torch::autograd::python::PythonEngine::getDefaultEngine();
+  return engine.execute(fmap(outputs, get_edge), grad_outputs, true, false, fmap(inputs, get_edge));
+}
+
+void assertAllClose(const tensor_list& a, const tensor_list& b) {
+  JIT_ASSERT(a.size() == b.size());
+  for (std::size_t i = 0; i < a.size(); ++i) {
+    JIT_ASSERT(a[i].is_same_size(b[i]));
+    JIT_ASSERT(a[i].allclose(b[i]));
+  }
+}
+
+void testAD() {
+  static const auto unwrap = [](const Variable& v) { return v.data(); };
+
+  using VL = variable_list;
+  static const var_meta_list binary_pointwise = {{2, 3, 4, 5}, {2, 3, 4, 5}};
+  static const std::vector<ADTestSpec> ad_tests = {
+    {"add", binary_pointwise, [](const VL& v) -> VL { return {v[0] + v[1]}; }},
+    {"sub", binary_pointwise, [](const VL& v) -> VL { return {v[0] - v[1]}; }},
+    {"mul", binary_pointwise, [](const VL& v) -> VL { return {v[0] * v[1]}; }},
+  };
+
+  // We have to release the GIL inside this method, because if we happen to
+  // initialize the autograd engine here, the newly spawned worker threads will
+  // try to initialize their PyThreadState*, and they need the GIL for this.
+  AutoNoGIL _no_gil;
+  for (const auto & test : ad_tests) {
+    // Get reference values form autograd
+    auto vars_in        = test.make_vars();
+    auto vars_out       = test(vars_in);
+    auto var_grads_in   = get_grad_outputs(vars_out);
+    auto var_grads_out  = grad(vars_out, vars_in, var_grads_in);
+
+    // Trace and differentiate the op
+    auto graph = trace(test, vars_in);
+    differentiate(graph);
+    Code bytecode {graph};
+    InterpreterState interpreter {bytecode};
+
+    // Get outputs from the interpreter
+    auto tensors_in                = fmap(vars_in, unwrap);
+    auto tensor_grads_in           = fmap(var_grads_in, unwrap);
+    tensor_list tensors_out, tensor_grads_out;
+    interpreter.runOneStage(tensors_in, tensors_out);
+    interpreter.runOneStage(tensor_grads_in, tensor_grads_out);
+
+    // Compare results
+    auto expected_tensors_out      = fmap(vars_out, unwrap);
+    auto expected_tensor_grads_out = fmap(var_grads_out, unwrap);
+    assertAllClose(tensors_out,      expected_tensors_out);
+    assertAllClose(tensor_grads_out, expected_tensor_grads_out);
+  }
+}
+
+std::string toString(std::shared_ptr<Graph>& graph) {
+  std::ostringstream s;
+  s << *graph;
+  return s.str();
+}
+
+void testGraphDiff() {
```

Comment:
We will need more rigorous testing for boundary cases, but that can wait until we have the meta-data for gradient, so that we can actually test by running the code.

Reply:
Yeah that was mostly meant as a smoke test so you can see what the output is